runtime/src/main/java/org/corfudb/runtime/LogReplicationRoutingQueueListener.java

```diff
@@ -0,0 +1,198 @@
+
+        List<CorfuStreamEntry> replicationStatusTableEntries =
+                entries.entrySet().stream().filter(e -> e.getKey().getTableName().equals(REPLICATION_STATUS_TABLE_NAME))
+                        .map(Map.Entry::getValue)
+                        .findFirst()
+                        .get();
+
+        for (CorfuStreamEntry entry : replicationStatusTableEntries) {
+
+            LogReplication.LogReplicationSession session = (LogReplication.LogReplicationSession)entry.getKey();
+
+            // Only process updates where operation type == UPDATE, model == Routing Queues and the client name
+            // matches
+            if (entry.getOperation() == CorfuStreamEntry.OperationType.UPDATE &&
+                    session.getSubscriber().getModel().equals(LogReplication.ReplicationModel.ROUTING_QUEUES) &&
+                    Objects.equals(session.getSubscriber().getClientName(), getClientName())) {
+                LogReplication.ReplicationStatus status = (LogReplication.ReplicationStatus)entry.getPayload();
+
+                if (status.getSinkStatus().getDataConsistent()) {
+                    // getDataConsistent() == true means that snapshot sync has ended.
+                    if (snapshotSyncInProgress.get()) {
+                        if (clientFullSyncPending.get()) {
+                            // Snapshot sync which was ongoing when the listener was subscribed has ended.  Attempt to
+                            // perform a full sync now.
+                            // TODO: Trigger full sync
+                            //  LogReplicationUtils.attemptClientFullSync(corfuStore, this, namespace);
+                            return;
+                        }
+                        // Process snapshot sync completion in steady state, i.e., client full sync is already complete
+                        snapshotSyncInProgress.set(false);
+                        onSnapshotSyncComplete();
+                    }
+                } else {
+                    // getDataConsistent() == false.  Snapshot sync has started.
+                    snapshotSyncInProgress.set(true);
+                    onSnapshotSyncStart();
+                }
+            }
+        }
+    }
+
+    //      -------- Methods to be implemented on the client/application  ---------------
+
+    /**
+     * Invoked when a snapshot sync start has been detected.
+     */
+    protected abstract void onSnapshotSyncStart();
+
+    /**
+     * Invoked when an ongoing snapshot sync completes
+     */
+    protected abstract void onSnapshotSyncComplete();
+
+    /**
+     * Invoked when data updates are received during a snapshot sync.  These updates will be the writes
+     * received as part of the snapshot sync
+     * @param results Entries received in a single transaction as part of a snapshot sync
+     */
+    protected abstract void processUpdatesInSnapshotSync(CorfuStreamEntries results);
```

Comment:
are you planning to change the return types in this PR?

Reply:
There are multiple changes needed on top of this. If the UT's are green, I think we can check-in this skeleton code and add on top as well.

Let me know if you suggest to change this PR and include all other items in this pr only.